dask/order.py

```diff
@@ -301,7 +301,20 @@ def finish_now_key(x):
                             num_needed[dep] -= 1
                         if not inner_stack:
                             if len(deps) == 1:
-                                inner_stack.append(dep)
+                                # An interesting option is to always set `inner_stack = [dep]` here.
```

Comment:
I was also looking into the failing map_overlap test. FWIW, here's what I cam up with to get things passing.

```diff
diff --git a/dask/order.py b/dask/order.py
index cd5e5799f..19c41f443 100644
--- a/dask/order.py
+++ b/dask/order.py
@@ -297,12 +297,19 @@ def order(dsk, dependencies=None):
                                 i += 1
 
                     if deps:
+                        num_needed_inner = []
                         for dep in deps:
                             num_needed[dep] -= 1
+                            if num_needed[dep] == 1:
+                                num_needed_inner.append(dep)
                         if not inner_stack:
                             if len(deps) == 1:
                                 inner_stack.append(dep)
                                 continue
+                            elif num_needed_inner:
+                                deps2 = sorted(deps, key=lambda x: -num_needed[x])
+                                inner_stack.extend(deps2)
+                                continue
                         else:
                             deps.discard(inner_stack[-1])  # safe to mutate
                             if not deps:
diff --git a/dask/tests/test_order.py b/dask/tests/test_order.py
index 98e450f48..80bac1a3b 100644
--- a/dask/tests/test_order.py
+++ b/dask/tests/test_order.py
@@ -516,7 +516,7 @@ def test_map_overlap(abcde):
        |/  | \ | / | \|
       d1  d2  d3  d4  d5
        |       |      |
-      e1      e2      e5
+      e1      e3      e5
 
     Want to finish b1 before we start on e5
     """
```

Your way is probably better, but wanted to put this up just in case.

I *think* the intuition here is "are we really close to being done? If so, let's prioritize these deps". Hopefully we're both achieving that.

Reply:
Neat!  I'll take a look tomorrow.  It's fun having somebody else poke at this.